src/types/list.rs

```diff
@@ -625,6 +704,26 @@ impl<'py> Iterator for BoundListIterator<'py> {
         }
     }
 
+    #[inline]
+    #[cfg(feature = "nightly")]
+    fn nth(&mut self, n: usize) -> Option<Self::Item> {
+        #[cfg(not(Py_LIMITED_API))]
+        {
+            self.with_critical_section(|index, length, list| unsafe {
+                Self::nth_unchecked(index, length, list, n)
+            })
+        }
+        #[cfg(Py_LIMITED_API)]
+        {
+            let Self {
+                index,
+                length,
+                list,
+            } = self;
+            Self::nth(index, length, list, n)
+        }
```

Comment:
```suggestion
        self.with_critical_section(|index, length, list| unsafe {
            Self::nth(index, length, list, n)
        })
```

If you implement my suggestion to only have `BoundListIterator::nth` then you can simplify this a lot.

Reply:
and you can do a similar refactor for the `nth_back` implementation for the `impl DoubleEndedIterator for BoundListIterator` block below